### ISSUE TITLE
The edges of the island aren't biome applied.

### ISSUE THREAD

**Issue by variananora** [BentoBoxWorld/Biomes#28]
#### Description
When selecting the biome it changes everything expect the edges of the island.

####  Steps to reproduce the behavior:
1. Do /is biomes
2. Select anything except Plains
3. Reload the chunk
4. Check the edges

#### Expected behavior
When selecting the biome it changes everything expect the edges of the island. 

#### Screenshots or video
https://youtu.be/zVUN8lbsI3Y

#### Server Information:
Settings used: 
  distance-between-islands: 128
  protection-range: 50
using permission bskyblock.island.range.125

 - Database being used (YAML, JSON, MySQL, MongoDB): YAML
 - OS: Debian 8.10
 - Java Version: Java 11
 - BentoBox version: 1.4.0
 - Running PAPER 1.13.2 - 624.
 - BentoBox version: 1.4.0
 - Loaded Game Worlds:
 - BSkyBlock_world (BSkyBlock)
 - Loaded Addons:
 - Biomes 0.4.5
 - BSkyBlock 1.4.2
 - Challenges 0.6.5-SNAPSHOT-#145
 - Greenhouse 0.0.2-SNAPSHOT
 - Level 1.4.0
 - Limits 0.2.0-SNAPSHOT
 - WelcomeWarps 1.4.0

**Comment by BONNe** [comment]
Can you check if this is still an issue in the latest development build?

**Comment by variananora** [comment]
Where do I get the dev build?
Is it the 4.5 Alpha release or it's the one that supports BBox 1.5.1?

**Comment by BONNe** [comment]
https://ci.codemc.org/job/BentoBoxWorld/job/Biomes/

**Comment by variananora** [comment]
Sorry for the late reply. It seems like it haven't been fixed yet.
/bbox v output:
Running PAPER 1.13.2.
BentoBox version: 1.5.1
Loaded Game Worlds:
bskyblock_world (BSkyBlock): Overworld, Nether, End
Loaded Addons:
Biomes 1.5.0.0 (ENABLED)
BSkyBlock 1.5.0 (ENABLED)
Challenges 0.7.5 (ENABLED)
Level 1.5.0 (ENABLED)
Limits 0.2.1-SNAPSHOT (ENABLED)
VoidPortals 1.5.0.0 (ENABLED)
WelcomeWarps 1.5.0 (ENABLED)

**Comment by BONNe** [comment]
Could you provide log file?

**Comment by variananora** [comment]
Is this it? (https://pastebin.com/1xV108uJ)

**Comment by BONNe** [comment]
Yea... that's weird... it shows that you have changed biome on an island with center X:0 Z:512
but you write that distance is 128, but here it looks like distance is 256. 

But in all situations, it changes biome on the whole island (the same coordinates), so that's good. 

Are you sure you have set 128 distance between islands? (and can you give coordinate of bedrock block for this island)

**Comment by variananora** [comment]
Oh, my bad. It is 256. Sorry for the miss-info there.
./bsb info output: https://pastebin.com/4wDTpRUC

**Comment by BONNe** [comment]
So, biome was applied correctly, from one corner to another.

I think what could happen, is that you are near spawn chunks, so probably these chunks do not refresh their biome, unless you restart game?

**Comment by variananora** [comment]
I tried to restart the Minecraft, and the chunk haven't updated yet.

**Comment by variananora** [comment]
Let me try reinstalling BentoBox

**Comment by variananora** [comment]
@BONNe It's still broken
https://youtu.be/2UrxyPG-Jj0
distance between island: 128
protection: 125

**Comment by BONNe** [comment]
Yea.. I found the issue.

Currently, Biomes addon cannot load chunks. So it is able to change biome only on loaded chunks.
In my server config, the range is 96x96 so it is 6 chunks radius.

I will check what  I can do there.

**Comment by BONNe** [comment]
Currently, biome updating is done in async task, to avoid game lagging. It prevents from loading new chunks in Spigot, so I am not able to load chunks to change biome on them in this situation.

So there are 3 solutions:

1. Load all chunks where biome must be updated before ASYNC task. And hope that they will not be unloaded while biome is changing.
2. Remove ASYNC task and update biome per chunk, but limit chunk count per tick. For example, each game tick only 1 chunk can change biome. So island with 128 radius will take 256 ticks.
3. Use the same ASYNC task as it is now, but collect and query unloaded chunks, and update biome when they are loaded.

Or is it more suggestions?

P.S. Changing biome is possible only in loaded chunks.

**Comment by tastybento** [comment]
@BONNe Sorry for the log delay. 

Have the async task still. It should know all the chunks it needs to change right? Then [query the chunk to see if it is loaded or not](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/Chunk.html#isLoaded--). If not load it, change the biome for all 16x16 blocks and then if it was previously unloaded,  [unload and save it](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/Chunk.html#unload-boolean-). I think using the unload(true) might be required to save the chunk. I believe that the server will not unload a chunk if there's a player around.

By this method, you'll keep loaded chunk counts low.

**Comment by BONNe** [comment]
The issue is that I cannot do it from async task. Chunk loading/unloading can be done in the main thread only.

Even if I query these chunks, I still need to process them in the main thread :(

**Comment by tastybento** [comment]
Okay, by async I actually meant pseudo-async. i.e., a Bukkit repeating task. It does a chunk or a few chunks per tick and works through a queue. Set up the queue for each island. Also, if you want to survive server restarts mid-chunk changes, then you could save the island biome in a table and clear it only when it's done.

Pseudo code:
```
// Save the Biome setting in progress flag
set inProgressBiomeSetting database (island, biome);

// Queue is a queue filled with all the chunks that need to set for this island
queue = new ConcurrentLinkedQueue<>() of (world, x, z, biome)
// Constant
MAX_CHUNKS_PER_TICK = 10
// Repeating task with field
BukkitTaskField = bukkit.repeating( task {
  count = 0;
  while( queue != empty  && count < MAX_CHUNKS_PER_TICK) {
    queue.poll {
          bool unloaded;
          if (world.getChunk(x,z).isUnloaded()) {
             unloaded = true;
             world.getChunk(x,z).loadchunk();
          }
          world.getChunk(x,z).setBiome(biome);
          if (unloaded) world.getChunk(x,z).unload(true);
          count++;
        }
  }
  if (queue.isEmpty) BukkitTaskField.cancel();
}, 0L, 1L})
```

something like that anyway

**Comment by BONNe** [comment]
Implemented in [Build #126](https://ci.codemc.org/job/BentoBoxWorld/job/Biomes/126/)
